pkg/retriever/retrieve.go

```diff
@@ -0,0 +1,438 @@
+	}
+
+	if len(candidates) == 0 {
+		return nil, ErrNoCandidates
+	}
+
+	acceptableCandidates := make([]RetrievalCandidate, 0)
+	for _, candidate := range candidates {
+		if cfg.IsAcceptableStorageProvider == nil || cfg.IsAcceptableStorageProvider(candidate.MinerPeer.ID) {
+			acceptableCandidates = append(acceptableCandidates, candidate)
+		}
+	}
+
+	if cfg.Instrumentation != nil {
+		if err := cfg.Instrumentation.OnRetrievalCandidatesFiltered(len(acceptableCandidates)); err != nil {
+			return nil, err
+		}
+	}
+
+	if len(acceptableCandidates) == 0 {
+		return nil, ErrNoCandidates
+	}
+
+	return acceptableCandidates, nil
+}
+
+// collectResults is responsible for receiving query errors, retrieval errors
+// and retrieval results and aggregating into an appropriate return of either
+// a complete RetrievalStats or an bundled multi-error
+func collectResults(ctx context.Context, retrieval *retrieval, candidates []RetrievalCandidate) (*RetrievalStats, error) {
+	var finishedCount int
+	var queryErrors error
+	var retrievalErrors error
+
+	for {
+		select {
+		case result := <-retrieval.resultChan:
+			if result.Err != nil {
+				if errors.Is(result.Err, ErrQueryFailed) {
+					queryErrors = multierr.Append(queryErrors, result.Err)
+				} else {
+					retrievalErrors = multierr.Append(retrievalErrors, result.Err)
+				}
+			}
+			if result.RetrievalStats != nil {
+				return result.RetrievalStats, nil
+			}
+			// have we got all responses but no success?
+			finishedCount++
+			if finishedCount >= len(candidates) {
+				if retrievalErrors == nil {
+					// we failed, but didn't get any retrieval errors, so must have only got query errors
+					retrievalErrors = ErrAllQueriesFailed
+				} else {
+					// we failed, and got only retrieval errors
+					retrievalErrors = multierr.Append(retrievalErrors, ErrAllRetrievalsFailed)
+				}
+				return nil, multierr.Append(queryErrors, retrievalErrors)
+			}
+		case <-ctx.Done():
+			return nil, context.Canceled
+		}
+	}
+}
+
+// runRetrievalCandidate is a singular CID:SP retrieval, expected to be run in a goroutine
+// and coordinate with other candidate retrievals to block after query phase and
+// only attempt one retrieval-proper at a time.
+func runRetrievalCandidate(ctx context.Context, cfg *RetrievalConfig, client RetrievalClient, retrieval *retrieval, candidate RetrievalCandidate) {
+	var timeout time.Duration
+	if cfg.GetStorageProviderTimeout != nil {
+		timeout = cfg.GetStorageProviderTimeout(candidate.MinerPeer.ID)
+	}
+
+	var stats *RetrievalStats
+	var done func()
+
+	// run the query phase
+	queryResponse, err := queryPhase(ctx, cfg, client, timeout, candidate)
+
+	if queryResponse != nil {
+		// if query is successful, then wait for priority and execute retrieval
+		done = retrieval.waitQueue.Wait(queryResponse)
+		if retrieval.canSendResult() {
+			stats, err = retrievalPhase(ctx, cfg, client, timeout, candidate, queryResponse)
+		}
```

Comment:
As much as it pains me to have the sender goroutine call close, I like the way this ended up _significantly_ more than the previous iterations because it reads much better. Calling the functions by phases helped as well I think.